### PR TITLE
Support Bucket Shuffle Join

### PR DESCRIPTION
**Pull request by HappenLee**
issue:#4394

## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] Documentation Update (if none of the other choices apply)
- [] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have create an issue on (Fix #ISSUE), and have described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If this change need a document change, I have updated the document
- [x] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/data_stream_sender.cpp [hunk 1]
```diff
@@ -479,7 +495,30 @@ Status DataStreamSender::send(RuntimeState* state, RowBatch* batch) {
...
+
+            for (auto ctx : _partition_expr_ctxs) {
+                void* partition_val = ctx->get_value(row);
+                // We can't use the crc hash function here because it does not result
+                // in uncorrelated hashes with different seeds.  Instead we must use
+                // fvn hash.
+                // TODO: fix crc hash/GetHashValue()
+                //hash_val = RawValue::get_hash_value_fvn(
+                //    partition_val, ctx->root()->type(), hash_val);
+                hash_val = RawValue::zlib_crc32(
```
Comment (kangkaisen) [inline be/src/runtime/data_stream_sender.cpp:517]: Why comment is fvn hash but code is crc32 hash?

File: fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java [hunk 2]
```diff
@@ -498,6 +520,72 @@ private boolean canColocateJoin(HashJoinNode node, PlanFragment leftChildFragmen
         return false;
     }
 
+    private boolean canBucketShuffleJoin(HashJoinNode node, PlanFragment leftChildFragment,
+                                   List<Expr> rhsHashExprs) {
+        if (!ConnectContext.get().getSessionVariable().isEnableBucketShuffleJoin()) {
+            return false;
+        }
+        // If user have a join hint to use proper way of join, can not be colocate join
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java:528]:
```suggestion
        // If user have a join hint to use proper way of join, can not be bucket shuffle join
```

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 3]
```diff
@@ -783,13 +784,39 @@ private void computeFragmentExecParams() throws Exception {
...
             }
 
-            // add destination host to this fragment's destination
-            for (int j = 0; j < destParams.instanceExecParams.size(); ++j) {
-                TPlanFragmentDestination dest = new TPlanFragmentDestination();
-                dest.fragment_instance_id = destParams.instanceExecParams.get(j).instanceId;
-                dest.server = toRpcHost(destParams.instanceExecParams.get(j).host);
-                dest.setBrpcServer(toBrpcHost(destParams.instanceExecParams.get(j).host));
-                params.destinations.add(dest);
+            if (bucketShuffleJoinController.isBucketShuffleJoin(destFragment.getFragmentId().asInt())) {
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:787]: If you could abstract the query schedule strategy like `Presto` and refactor this class, which would be very great.

File: be/src/runtime/data_stream_sender.cpp (hunk 1 above)
Comment (HappenLee) [reply be/src/runtime/data_stream_sender.cpp:517]: This comment is not right， It should be deleted.

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java (hunk 3 above)
Comment (HappenLee) [reply fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:787]: Yes, It is the future work to do. But this part is complex and has a great impact. I think we need to this work gradually

File: be/src/runtime/data_stream_sender.cpp (hunk 1 above)
Comment (HappenLee) [reply be/src/runtime/data_stream_sender.cpp:517]: This comment is not right， I will change this part.

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java (hunk 3 above)
Comment (kangkaisen) [reply fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:787]: OK

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**